### DISCUSSION TITLE
LOG and printf do not print negative values when compile to run on devices

### DISCUSSION THREAD

**Discussion by rahav** [contiki-ng/contiki-ng#2728] [Q&A] [answered]
A statement such as `printf("Hello, world. i = %i, d = %d \n", val, val);` will print negative value if `int val` but not `char val`. Is that the expected behavior?

**Comment by pjonsson** [comment]
If it's an ARM-platform, you can try compiling with `make MAKE_WITH_LIB_DBG_IO=0 ...` to use the libc shipped with your ARM SDK, and see if that makes any difference. You probably need to make clean before though, I doubt the build system will detect changed flags and rebuild the dependencies.

As a rule of thumb, the default printf in Contiki-NG should behave like any other C99-compliant printf. The documented exception that I'm aware of is floating point in dbg-io, which is silently removed. Even the dbg-io implementation is too large for the MSP430 platforms, so that has yet another printf implementation which is limited in different ways.

**Reply by rahav** [reply]
The problem appears on nRF52840. @pjonsson I think that it may be a binary representation issue. I started to suspect that because, for example, it prints 253 when the value is -3. I tested it using the slightly modified `hello-world`, the code is below.

Unless I am missing something really naive, it appears that there is a bigger problem. The **internal representation is not equal to a CONSTANT that holds the same value**, when the TARGET is `nrf52840`. I tried the code I included below on `sky` and it runs correctly. I do not have other ARM devices, so I do not know if this is a more general problem or limited to the nrf. Here is what an nrf mote will output:

> val     : i = 253, d = 253 
intermed: i = 253, d = 253 
val = 253 == intermed = 253
val = 253 == val2 = 253
val = 253 != -3
val = 253 != -3
intermed = 253 != -3
intermed = 253 != -3

There will be no such problem when compiled for `.native`, or `sky`.

Here is the code
````
#include "contiki.h"

#include <stdio.h> /* For printf() */

#define NEG3 -3

/*---------------------------------------------------------------------------*/
PROCESS(hello_world_process, "Hello world process");
AUTOSTART_PROCESSES(&hello_world_process);
/*---------------------------------------------------------------------------*/
PROCESS_THREAD(hello_world_process, ev, data)
{
  static struct etimer timer;

  static char val, val2;
  static int intermed;

  PROCESS_BEGIN();

  val = val2 = -3;

  /* Setup a periodic timer that expires after 10 seconds. */
  etimer_set(&timer, CLOCK_SECOND * 2);

  while(1) {
    intermed = val;
    printf("val     : i = %i, d = %d \n", val, val);
    printf("intermed: i = %i, d = %d \n", intermed, intermed);

    if (val == intermed) {
      printf("val = %d == intermed = %d\n", val, intermed);
    } else {
      printf("val = %d != intermed = %d\n", val, intermed);
    }

    if (val == val2) {
      printf("val = %d == val2 = %d\n", val, val2);
    } else {
      printf("val = %d != val2 = %d\n", val, val2);
    }

    if (val == -3) {
      printf("val = %d == %d\n", val, -3);
    } else {
      printf("val = %d != %d\n", val, -3);
    }

    if (val == NEG3) {
      printf("val = %d == %d\n", val, NEG3);
    } else {
      printf("val = %d != %d\n", val, NEG3);
    }

    if (intermed == -3) {
      printf("intermed = %d == %d\n", val, -3);
    } else {
      printf("intermed = %d != %d\n", val, -3);
    }

    if (intermed == NEG3) {
      printf("intermed = %d == %d\n", val, NEG3);
    } else {
      printf("intermed = %d != %d\n", val, NEG3);
    }

    printf("\n");

    /* Wait for the periodic timer to expire and then restart the timer. */
    PROCESS_WAIT_EVENT_UNTIL(etimer_expired(&timer));
    etimer_reset(&timer);
  }

  PROCESS_END();
}
````

To be complete, when I added the suggested parameter: `make TARGET=nrf52840 BOARD=dk PORT=/dev/ttyACM0 MAKE_WITH_LIB_DBG_IO=0 hello-world.upload` it outputted nothing in all attempts but one. Nevertheless, in that once, it did not print it correctly.

Thank you,
Rahav

**Comment by alexrayne** [comment]
I got sometime similar troubles , fixing by extlisit typecasting args to
types declared by format specifiers

ср, 4 окт. 2023 г., 18:17 Rahav Dor ***@***.***>:

**Reply by rahav** [reply]
Casting does not work for me. Can you post an example please?

But even if it works for printing, this is a major problem for control systems. Imagine what can happen when level gets to be -1, and following condition does not work:
````
STOP_HEATING -1
int level;

if (level == STOP_HEATING)
````

**Reply by pjonsson** [reply, marked as answer]
Char is not guaranteed to be signed/unsigned: https://stackoverflow.com/questions/2054939/is-char-signed-or-unsigned-by-default.

Use an `int8_t` if you need a small counter, as a bonus that will give `-1` as value when you print the variable in gdb.
